contrib/check-for-api-changes.sh

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env bash
+#
+# Checks the public API of crates, exits with non-zero if there are currently
+# changes to the public API not already committed to in the various api.txt
+# files.
+
+set -e
+
+# Crates in this repo to check.
+CRATES="bitcoin hashes"
+
+export RUSTDOCFLAGS='-A rustdoc::broken-intra-doc-links'
+
+main() {
+    local repo_dir=$(git rev-parse --show-toplevel)
+
+    check_we_have_nightly
+
+    if is_dirty ; then
+        echo "Cannot check for API changes with a dirty index, commit your changes first"
+        exit 1
+    fi
+
+    pushd "$repo_dir/bitcoin" > /dev/null
+    cargo public-api --no-default-features --features=no-std > "../api/bitcoin/api-no-std.txt"
+    cargo public-api > "../api/bitcoin/api-default-features.txt"
+    cargo public-api --all-features > "../api/bitcoin/api-all-features.txt"
+    popd > /dev/null
+
+    pushd "$repo_dir/hashes" > /dev/null
+    cargo public-api --no-default-features > "../api/hashes/api-no-features.txt"
+    cargo public-api > "../api/hashes/api-default-features.txt"
+    cargo public-api --no-default-features --features=alloc > "../api/hashes/api-alloc.txt"
+    cargo public-api --all-features > "../api/hashes/api-all-features.txt"
+    popd > /dev/null
+
+    if is_dirty ; then
+        echo "You have introduced changes to the public API, commit the changes currently in your working directory"
+        exit 1
+    else
+        echo "No changes to the current public API"
+    fi
+}
+
+# Returns true if the working directory is dirty.
+is_dirty() {
+    if git diff --quiet; then
```

Comment:
In a3dde3e716413a0606540fbf9c452a6bc63c4086:

I think we should actually only run `git diff` on the `api/` directory. (You need to promote `repo_dir` to a global variable to get the paths straight.)